mlir/lib/Dialect/Affine/IR/AffineOps.cpp

```diff
@@ -1947,11 +1947,183 @@ struct AffineForEmptyLoopFolder : public OpRewritePattern<AffineForOp> {
     return success();
   }
 };
+
+struct AffineForReductionIter : public OpRewritePattern<AffineForOp> {
+  using OpRewritePattern<AffineForOp>::OpRewritePattern;
+
+  bool isInCurrentAffineFor(Operation *op, AffineForOp forOp) const {
+    auto *curOp = op;
+    while (auto *parentOp = curOp->getParentOp()) {
+      auto maybeParentFor = dyn_cast<AffineForOp>(parentOp);
+      if (maybeParentFor && maybeParentFor == forOp)
+        return true;
+      curOp = parentOp;
+    }
+    return false;
+  }
+
+  bool hasAllDimsReduced(ArrayRef<Value> indices, Value indVar) const {
+    if (llvm::all_of(indices,
+                     [indVar](Value index) { return index != indVar; }))
+      return true;
+    return false;
+  }
+
+  // bool isAffineLoadWithReducedDims(Operation *op, Value indVar) const {
+  //  if (auto load = dyn_cast<AffineLoadOp>(op)) {
+  //    SmallVector<Value, 4> indices(load.indices());
+  //    return hasAllDimReduced(indices, indVar);
+  //  }
+  //  return false;
+  //}
+
+  // bool isAffineStoreWithReducedDims(Operation *op, Value indVar) const {
+  //  if (auto store = dyn_cast<AffineStoreOp>(op)) {
+  //    SmallVector<Value, 4> indices(store.indices());
+  //    return hasAllDimReduced(indices, indVar);
+  //  }
+  //  return false;
+  //}
+
+  template <typename T>
+  bool haveSameIndices(AffineLoadOp load, T storeOrLoad) const {
+    static_assert(llvm::is_one_of<T, AffineLoadOp, AffineStoreOp>::value,
+                  "applies to only AffineLoadOp or AffineStoreOp");
+    SmallVector<Value, 4> loadIndices(load.indices());
+    SmallVector<Value, 4> storeIndices = storeOrLoad.indices();
+    if (loadIndices.size() != storeIndices.size())
+      return false;
+    return std::equal(loadIndices.begin(), loadIndices.end(),
+                      storeIndices.begin());
+  }
+
+  // bool isCurrentLoad(Operation *op, AffineLoadOp load) const {
+  //  auto currentLoad = dyn_cast<AffineLoadOp>(op);
+  //  if (currentLoad && currentLoad == load)
+  //    return true;
+  //  return false;
+  //}
+
+  LogicalResult matchAndRewrite(AffineForOp forOp,
+                                PatternRewriter &rewriter) const override {
+    Block *block = forOp.getBody();
+    SmallVector<std::pair<Operation *, Operation *>, 0> candidateOpsInFor;
+    SmallVector<Operation *> candidateLoadsOutsideFor;
+    block->walk([&](Operation *operation) {
+      if (auto load = dyn_cast<AffineLoadOp>(operation)) {
+        Value memref = load.getMemRef();
+        bool foundLoad = false;
+        bool foundStore = false;
+        for (auto user : memref.getUsers()) {
+          if (auto maybeOutsideLoad = dyn_cast<AffineLoadOp>(user)) {
+            // check for a load outside the current for.
+            if (!foundLoad &&
+                !isInCurrentAffineFor(maybeOutsideLoad.getOperation(), forOp) &&
+                maybeOutsideLoad.getMemRef() == memref &&
+                haveSameIndices<AffineLoadOp>(load, maybeOutsideLoad)) {
```

Comment:
I'm not sure that a sameIndices check is sufficient here. For example imagine a call to a function that load/stores. Or a computation which may alias the given memory location.

Reply:
Do you have Alias analysis in mind? Alternatively, we could locate the alloca operation and check if all its users are the candidate load/store operations. In any case, do we need to handle partially reduced memref (i.e., only one dimension is the reduced one)? Currently, all the dimensions need to reduce.

Reply:
I think alias analysis needs to check that the only potential uses of are load/stores, and then of the load stores there aren't potentially aliasing load/stores that would conflict

Reply:
Until there is a real alias analysis, we can assume that (1) memrefs that are supplied as function arguments do not alias (the rest of affine infrastructure makes this assumption anyway) and (2) memrefs obtained by view/subview form mutually aliasing groups (even though views may be disjoint). With these assumptions, I would say we need to check that the operations between the load and the store are known not to access a memref that aliases the "reduction" memref using the side effects interface. For function calls specifically, we don't know anything at all, so we assume they access everything. Same for operations, including unknown/unregistered, that do not implement the interface -- conservatively bail out.